crates/ty_python_semantic/src/types.rs

```diff
@@ -7185,6 +7271,158 @@ impl<'db> FieldInstance<'db> {
+                NewTypeBase::ClassType(base_class.materialize(db, materialization_kind))
+            }
+            NewTypeBase::NewType(base_newtype) => {
+                NewTypeBase::NewType(base_newtype.materialize(db, materialization_kind))
+            }
+        };
+        Self::new(
+            db,
+            self.name(db).clone(),
+            self.definition(db),
+            materialized_base,
+        )
+    }
+
+    fn apply_type_mapping_impl<'a>(
+        self,
+        db: &'db dyn Db,
+        type_mapping: &TypeMapping<'a, 'db>,
+        visitor: &ApplyTypeMappingVisitor<'db>,
+    ) -> Self {
+        let mapped_base = match self.base(db) {
+            NewTypeBase::ClassType(base_class) => NewTypeBase::ClassType(
+                base_class.apply_type_mapping_impl(db, type_mapping, visitor),
+            ),
+            NewTypeBase::NewType(base_newtype) => NewTypeBase::NewType(
+                base_newtype.apply_type_mapping_impl(db, type_mapping, visitor),
+            ),
+        };
+        Self::new(db, self.name(db).clone(), self.definition(db), mapped_base)
+    }
+
+    fn is_subtype_of(self, db: &'db dyn Db, other: Self) -> bool {
+        let mut candidate_newtype = self;
+        loop {
+            if candidate_newtype == other {
+                // Either `self` is `other`, or it's a (transitive) NewType wrapper of it.
+                return true;
+            }
+            match candidate_newtype.base(db) {
+                NewTypeBase::NewType(base_newtype) => candidate_newtype = base_newtype,
+                // Classes can't inherit from NewTypes, so if we reach the base `ClassType` without
+                // seeing `other`, then we can't be a subtype of it.
+                NewTypeBase::ClassType(_) => return false,
+            }
+        }
+    }
+
+    fn has_relation_to_impl<C: Constraints<'db>>(
+        self,
+        db: &'db dyn Db,
+        other: Self,
+        _: TypeRelation,
+        _: &HasRelationToVisitor<'db, C>,
+    ) -> C {
+        // TODO: Is this correct?
+        C::from_bool(db, self.is_subtype_of(db, other))
+    }
+
+    pub(super) fn is_disjoint_from_impl<C: Constraints<'db>>(
+        self,
+        db: &'db dyn Db,
+        other: Self,
+        _: &IsDisjointVisitor<'db, C>,
+    ) -> C {
+        // Two NewTypes are disjoint if they're not equal and neither inherits from the other.
+        // (NewTypes have single inheritance, and a regular class can't inherit from a NewType, so
+        // it's not possible for some third type to multiply-inherit from both.)
+        C::from_bool(
+            db,
+            !self.is_subtype_of(db, other) && !other.is_subtype_of(db, self),
+        )
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, get_size2::GetSize)]
+pub enum NewTypeBase<'db> {
+    ClassType(ClassType<'db>),
+    NewType(NewTypeInstance<'db>),
+}
+
+impl<'db> NewTypeBase<'db> {
+    fn to_type(self, db: &'db dyn Db) -> Type<'db> {
```

Comment:
```suggestion
    fn to_instance(self, db: &'db dyn Db) -> Type<'db> {
```

to make it clear that this method creates the type that represents "the set containing all possible instances of this newtypebase", not the type "the singleton set that only contains the runtime object representing this newtypebase itself"